hazelcast/src/main/java/com/hazelcast/map/impl/proxy/MapProxySupport.java

```diff
@@ -716,12 +796,72 @@ protected void putAllInternal(Map<?, ?> map) {
+        return (putAllBatchSize > 0 && mapSize > putAllBatchSize * getNodeEngine().getClusterService().getSize());
+    }
+
+    @SuppressWarnings("checkstyle:magicnumber")
+    private int getPutAllInitialSize(boolean useBatching, int mapSize, int partitionCount) {
+        if (mapSize == 1) {
+            return 1;
+        }
+        if (useBatching) {
+            return putAllBatchSize;
+        }
+        if (putAllInitialSizeFactor < 1) {
+            // this is an educated guess for the initial size of the entries per partition, depending on the map size
+            return (int) ceil(20f * mapSize / partitionCount / log10(mapSize));
+        }
+        return (int) ceil(putAllInitialSizeFactor * mapSize / partitionCount);
+    }
+
+    private int getPutAllFutureSize(int mapSize, boolean useBatching, int partitionCount) {
+        return (useBatching ? round((float) partitionCount * mapSize / putAllBatchSize) : partitionCount);
+    }
+
+    private void invokePutAllOperation(Address address, List<Integer> memberPartitions, List<Future> futures,
+                                       MapEntries[] entriesPerPartition) {
+        int size = memberPartitions.size();
+        int[] partitions = new int[size];
+        int index = 0;
+        for (Integer partitionId : memberPartitions) {
+            if (entriesPerPartition[partitionId] != null) {
+                partitions[index++] = partitionId;
+            }
+        }
+        if (index == 0) {
+            return;
+        }
+        // trim partition array to real size
+        if (index < size) {
+            partitions = Arrays.copyOf(partitions, index);
+            size = index;
+        }
+
+        index = 0;
+        MapEntries[] entries = new MapEntries[size];
+        long totalSize = 0;
+        for (int partitionId : partitions) {
+            int batchSize = entriesPerPartition[partitionId].size();
+            assert (putAllBatchSize == 0 || batchSize <= putAllBatchSize);
+            entries[index++] = entriesPerPartition[partitionId];
+            totalSize += batchSize;
+            entriesPerPartition[partitionId] = null;
+        }
+        if (totalSize == 0) {
+            return;
+        }
+
+        Future future = createPutAllOperationFuture(name, totalSize, partitions, entries, address);
+        futures.add(future);
+    }
+
+    protected Future createPutAllOperationFuture(final String name, final long size, int[] partitions, MapEntries[] entries,
+                                                 Address address) {
+        MapOperation op = operationProvider.createPutAllPerMemberOperation(name, partitions, entries);
         final long time = System.currentTimeMillis();
-        InternalCompletableFuture<Object> future = operationService.invokeOnPartition(SERVICE_NAME, op, partitionId);
+        InternalCompletableFuture<Object> future = operationService.invokeOnTarget(SERVICE_NAME, op, address);
```

Comment:
Can it be more appropriate to use a logic similar to `invokeOnAllPartitions` here because it internally relies on `InvokeOnPartitions` and it has retry mechanism in case of failures ? Btw, I wrote a small test [here](https://gist.github.com/ahmetmircik/ad727c8ae6158a7a4be3e7fae933467e) and it fails on master and passes on maintenance.